pages/blog/json-schema-beyond-validation.md

```diff
@@ -0,0 +1,151 @@
+
+## Quick intro to JSON Schema
+
+Let's try and take a look at an example. Given the following, we have a schema representing the validation rules that our data should comply with. 
+
+<figure>
+  <img src="../img/posts/jonaslagoni/json-schema-beyond-validation/json-schema-process.webp" title="Processor setup"/>
+  <figcaption className="text-center text-gray-400 text-sm">Displays the overall process of validating data using JSON Schema.</figcaption>
+</figure>
+
+The JSON Schema defines that the JSON data should be an object, which requires a property called `someRequiredProperty` to always be present and an optional property called `someOptionalProperty`. `someRequiredProperty` should validate against an integer and and `someOptionalProperty` against an arbitrary string. The schema also dictates that no additional properties (`"additionalProperties": false`) may be allowed. There is also some metadata defined, called `$id` and `$schema`, but they are not important for this example.
+
+If we then take a look at the example [data instances](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-01#section-4.2), the first one contains the required property, and the second one both the required and the optional property.
+
+The data and the JSON Schema can then together, validate whether the data is an instance of the schema i.e. validate whether the data comply with the validation rules and giving a simple true or false statement whether they are compatible.
+
+This is an extremely powerful tool which allows you to create complex validation rules for data, that would not otherwise be easy to do, if you used other specification which defines the data model using [TypeSchema](https://typeschema.org/), [JTD](https://datatracker.ietf.org/doc/html/rfc8927), etc.
+
+## The problems of using JSON Schema as data definitions
+
+Many of the JSON Schema keywords are for [JSON instance validation](https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-validation-01), which means specifying validation rules that data should comply with. However, what if you wanted to know the definition of the data rather than what it should validate against? That is currently not something JSON Schema provides to you, even though it is such an important part of tooling. Let's deep dive a bit into JSON Schema and see where some of the complexity lies:
+
+It of course not all JSON Schemas that are complex to interpret to data definition. For our simple example, we can almost interpret it as is. If I wanted a class in TypeScript that represented the data, it could look something like this (gonna use TS syntax as examples throughout). Notice: how we use the `$id` keyword to define the name of such a class.
```

Comment:
```suggestion
Of course not all JSON Schemas are complex to interpret to data definition. For our simple example, we can almost interpret it as is. If I wanted a class in TypeScript that represented the data, it could look something like this (gonna use TS syntax as examples throughout). Notice: how we use the `$id` keyword to define the name of such a class.
```